### ISSUE TITLE
Negative numbers are not being parsed correctly inside records when directly next to =

### ISSUE THREAD

**Issue by mrakgr** [Microsoft/visualfsharp#3665]
```
type Q = {by: int}
{by=-1}
```
It is a small thing, but the above gives an error. To fix it, the -1 needs to be further apart from the equals character like in the following example.
```
type Q = {by: int}
{by= -1}
```
This is in F# 4.0. I am not sure if it was fixed in the latest version.

**Comment by realvictorprm** [comment]
Happens also in F# Interactive 4.1 :(

Thanks for the report @mrakgr 
I think this is a bug but I'm not sure.

**Comment by mrakgr** [comment]
It also happens in standard let statements such as `let a = -1`. I'd be interested in knowing whether this should be considered a bug too.

**Comment by mrakgr** [comment]
The main reason I am asking is because I am making a language with a syntax very similar to F#'s and I had these exact problems happen to me there, so I went ahead and fixed it just now. I think the grammar is fairly unambiguous in record assignments and let statements so it should not be a problem to remove the necessity of having a whitespace character between `=` and `-`

**Comment by realvictorprm** [comment]
@mrakgr problem is here that it's threat as operator (which isn't that wrong)

**Comment by dsyme** [comment]
It's by design - '=-' is an operator token

**Comment by mrakgr** [comment]
I am not too familiar with lexers as I am using Fparsec for my language, but it was an easy fix for me. I am guessing that since F# is tokenizing operators before the later passes, this would be a very difficult thing to change? 

If so, I am happy to let this issue rest.

**Comment by forki** [comment]
=- is a valid operator name. So you can "fix it" without making that
operator invalid. That would be a breaking change since people might
already use it.

Am 29.09.2017 18:11 schrieb "mrakgr" <notifications@github.com>:

**Comment by mrakgr** [comment]
No, I did not mean suggest that =- should be made invalid. What I did in  `let ... =` and `{...=...; ...=...}` is parse the `=` without considering them as operators, rather they are just pieces of syntax. This was easy to do as I am using parser combinators, but if you are doing tokenization then things are more difficult. I suppose you could try splitting the `=-` token into two in the later stage.

**Comment by forki** [comment]
Or we keep like now and don't introduce new ambiguities ;-)

Am 29.09.2017 18:21 schrieb "mrakgr" <notifications@github.com>:

**Comment by mrakgr** [comment]
It would not introduce ambiguities right now, as far as I can tell. The only thing it might do is block introducing new let forms with different separator keywords, such as `let ... =-`, `let ... =+` or something along those lines. It would only really concern the unary operators immediately following the `=` in let and record statements.

**Comment by forki** [comment]
Ok. Let me ask more constructively: what do you think is the gain here?

Am 29.09.2017 18:38 schrieb "mrakgr" <notifications@github.com>:

**Comment by mrakgr** [comment]
A slight syntax improvement for the language.

**Comment by forki** [comment]
let x = ref false
let y = { A=!x}


Should work as well? And all operators combined with =?

What about {A=-a=a}?

That will bring interest new ways to improve our error reporting :-)

Am 29.09.2017 18:45 schrieb "mrakgr" <notifications@github.com>:

A slight syntax improvement for the language.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/visualfsharp/issues/3665#issuecomment-333177831>,
or mute the thread

**Comment by mrakgr** [comment]
Ok, I give. 

I think that to make a better argument for my case here I would need to look into the F# compiler and see what is possible and isn't during the parsing stage, and how it interacts with error reporting. Not something I am interested in doing right now, at any rate.

**Comment by forki** [comment]
Sorry. I didn't want to scare you away. But this is really a very difficult
change. Maybe not impossible, but very difficult. And a lot can go wrong.

Am 29.09.2017 18:58 schrieb "mrakgr" <notifications@github.com>:

**Comment by abelbraaksma** [comment]
Note that this is actually quite consistent. Operators are roughly defined by a set of characters that can be combined to form new operators. Chaining operators together without spaces in-between them is not possible and is not only ambiguous, it non-deterministic:

Suppose you have a hypothetical (but valid!) operator like so:

```f#
let (=-) x y = -x
let foo = 12 =- 0
```

Now, you could argue, did I write `12 = -0` (i.e, is `foo` set to `false`) or did I write `(=-) 12 0` (i.e., is `foo` set to `-12`)?

_If_ a fix like this would be applied, you'd expect the compiler to know the difference, or raise perhaps a warning or error saying that the operator resolution is ambiguous. 

You could, of course, also argue that this fix should only apply to assignments, but assignment-operators are, from the point of view of F#, also "just operators", and they are essentially patterns, so any fix here has a much wider application.

An on another point of view, it is not unlikely that `=-` and `=+` will, at some point, become shorthand syntax just like in C# for `x = x - 123`, which wouldn't be possible given a "fix" in this regard.

But currently, even though perhaps not perfect, at least it is consistent. All of these are invalid:

```f#
let x=!foo
let y=-10
let z=+10
```

Any of these will raise a `Unexpected infix operator` exception.

So, to summarize, while I sympathize with your idea, and I have been struck by being confused by it in the past, even if we _wanted_ to fix this, I don't think there's a viable, deterministic, forward and backwards compatible solution possible.

**Comment by mrakgr** [comment]
After doing more experimentation on my own language which has very similar syntax to F#, I came to the conclusion that `let ... =-x` would indeed be ambiguous, but not for the reasons stated above. As a background, F# has fairly complex way of disambiguating the unary from binary operators.

For example, in `let f g x = g -x`, the `g` would get inferred as `int -> 'a` and hence the one `-` as a unary operator. In `let f g x = g - x`, now that I put in a whitespace, the `-` is a binary operator and g is typed as an `int`.

I thought the rule might be that whenever `-` is directly next to a variable it, but testing it a bit further, in `let f g x = g-x`, the `g` is also an `int`. So the parser checks the preceding character before deciding whether to interpret the operator as unary or binary. Testing it a bit more, in `let f g x = (g)-x` , `g` is also an int which shows that it is not just checking for letters or numbers, but probably separators such as start of file, newline, whitespace, `,`, `;` and parentheses.

Hence, a statement `let x=-1` would be ambiguous due to the unary operator failing the separator check for the preceding character.

The way I did it previously in my language is by having negate lower precedence than binary operators and application, but that is a mistake since statements such as `let f g x = g <| -x` would get parsed like `let f g x = g (-) x`. Once I fixed that today, I realized that I have in fact no way disambiguating `=-` from `=` and `-` even if the later pass could split the tokens – not without having to add state that remembers the position of the separator character when going into statements.

The reason why I am going into detail is to tie up the loose end. I've been coding in F# for 2 years now, but never bothered to think about the rules governing the parsing behavior of unary operators. Until I wrote them out just now, my mental model of what they were was wrong even if my feel for it was good enough.
